<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Robust MALDI mass spectra data pre-processing pipeline with MALDIrppa • MALDIrppa</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Robust MALDI mass spectra data pre-processing pipeline with MALDIrppa">
<meta property="og:description" content="MALDIrppa">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MALDIrppa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MALDIrppa_vignette.html">Robust MALDI mass spectra data pre-processing pipeline with MALDIrppa</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="MALDIrppa_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Robust MALDI mass spectra data pre-processing pipeline with MALDIrppa</h1>
                        <h4 class="author">
<a href="http://www.bioss.ac.uk/people/javier.html">Javier Palarea-Albaladejo</a>, Biomathematics &amp; Statistics Scotland, Edinburgh, UK</h4>
            
            <h4 class="date">2020-08-24</h4>
      
      
      <div class="hidden name"><code>MALDIrppa_vignette.Rmd</code></div>

    </div>

    
    
<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>
<p>To cite MALDIrppa in publications, please use:</p>
<p>Palarea-Albaladejo J., McLean K., Wright F. and Smith (2018). MALDIrppa: quality control and robust analysis for mass spectrometry data. Bioinformatics 34(3):522–523. (<a href="http://dx.doi.org/10.1093/bioinformatics/btx628" class="uri">http://dx.doi.org/10.1093/bioinformatics/btx628</a>).</p>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Mass spectrometry is actively being used to discover proteomic patterns in complex mixtures of peptides derived from biological samples. Reproducibility is however a significant challenge in MALDI protein profiling. Approaches to improve the analytical performance of MALDI protein profiling include extensive pre-fractionation strategies, immunocapture, pre-structured target surfaces, standardized matrix (co)crystallization, improved instrument components, internal standard peptides, quality-control samples, among others <span class="citation">(Albrethsen 2007)</span>. Mass spectrometry (MS) data pre-processing algorithms play a crucial role in rendering the subsequent data analysis more robust and accurate. The package <code>MALDIrppa</code> contributes a number of procedures for robust pre-processing and analysis, along with a number of functions to facilitate common data management operations. It is thought to work in conjunction with the <code>MALDIquant</code> package <span class="citation">(Gibb and Strimmer 2012)</span>, using object classes and methods from this latter. These notes provide basic guidelines for users to run a complete robust MALDI mass spectra pre-processing pipeline using their own data with small adaptations. Further specific details and examples on the full functionality of the <code>MALDIrppa</code> package can be found in the associated help pages.</p>
<p>The data set <code>spectra</code> used as example here is an anonymised collection of mass spectra trimmed to the [2500, 13000] <em>m/z</em> interval. Their <em>m/z</em> resolution is 30 times lower than the original (achieved using the <code>redResolution</code> function) so as to make it manageable for sharing and illustration purposes. It consists of 4 technical replicates of 5 biological replicates from 20 bacterial isolates. The mass spectra were originally acquired over the [2000, 20000] Daltons mass range using the Bruker Daltonics Ultraflex II mass spectrometer. The data set <code>type</code> contains a simplified version of the associated metadata.</p>
<p>The following sections go through the stages to transform raw mass spectra into formats suitable for downstream data analysis and modelling for biological interpretation. These involve data transformation, smoothing, baseline correction, normalisation, peak detection and peak alignment and binning. The features of the signals depend on technological progress and characteristics of the species under study. Unfortunately there is not a standard methodology for MS data pre-processing best working in all cases. This document illustrates a generic pipeline through <code>MALDIrppa</code> in combination with <code>MALDIquant</code> based on robust statistical methods which contributes to lessen the intrinsic reproducibility issues with MS data and facilitates obtaining reliable biological conclusions <span class="citation">(Mclean et al. 2017)</span>. However, the finest pre-processing in a particular case might require further discussion of choices for parameter settings and methods.</p>
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<p>The data are included in <code>MALDIrppa</code> in the own R’s <code>RData</code> binary file format. They can be loaded into the workspace using the <code>data</code> function.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MALDIrppa</span>)</pre></body></html></div>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">spectra</span>) <span class="co"># list of MassSpectra class objects</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">type</span>)    <span class="co"># metadata matrix</span></pre></body></html></div>
<p>Alternatively, raw data stored in text files can be easily imported into R and given the convenient <code>MassSpectrum</code> class format using the <code>importSpectra</code> function included in <code>MALDIrppa</code>. For importing data from more specialised file formats we suggest the reader to check out the package <code>MALDIquantForeign</code>. The function <code>summarySpectra</code> computes a table including some basic summary statistics. A look at these along with some graphical representations are useful to get a first insight into the data set and identify possible issues. The next code shows numerical results for the first 10 signals.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="../reference/summarySpectra.html">summarySpectra</a></span>(<span class="no">spectra</span>[<span class="fl">1</span>:<span class="fl">10</span>])</pre></body></html></div>
<pre><code>##           ID No.MZ  Min.MZ   Max.MZ Min.Int  Mean.Int  Std.Int Med.Int  MAD.Int Max.Int
## 1  160408F21  1857 2500.05 12995.77      19  722.2649 572.9375     664 486.2928    5888
## 2  160408F22  1857 2500.05 12995.77      18  637.1357 531.1853     582 382.5108    6128
## 3  160408F23  1857 2500.05 12995.77       8  542.4976 447.7237     483 401.7846    4713
## 4  160408F24  1857 2500.05 12995.77      34  957.3430 737.8402     903 587.1096    7843
## 5  160408G01  1857 2500.05 12995.77      42 1129.8185 871.5639    1064 652.3440    7860
## 6  160408G02  1857 2500.05 12995.77      24  637.1648 530.9029     595 358.7892    5459
## 7  160408G03  1857 2500.05 12995.77      12  505.9499 442.6567     466 298.0026    5198
## 8  160408G04  1857 2500.05 12995.77      43  984.1179 754.5525     936 530.7708    7333
## 9  160408G05  1857 2500.05 12995.77      27  906.9138 686.9536     854 558.9402    6068
## 10 160408G06  1857 2500.05 12995.77      46  906.9133 689.0228     891 481.8450    7505</code></pre>
<p>A number of parameters are required to be set for the different pre-processing stages. The functions include <em>safe</em> default values, but parameter settings throughout pre-processing stages interact in complex ways. Using simulation, we customised some of the key parameter so as to obtain optimal combinations in terms of sensibility and false discovery rate in peak profiles. They have been slightly modified to accommodate the characteristics of the reduced example data set used here. It is convenient to write an initial section in the pre-processing pipeline script to set the values for the parameters which will be used later on. This facilitates testing results from different settings.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co"># Pre-processing parameter settings</span>
<span class="no">thScale</span> <span class="kw">&lt;-</span> <span class="fl">2.5</span> <span class="co"># Smoothing</span>
<span class="no">ite</span> <span class="kw">&lt;-</span> <span class="fl">105</span> <span class="co"># Baseline correction</span>
<span class="no">SigNoi</span> <span class="kw">&lt;-</span> <span class="fl">2</span> <span class="co"># Peak extraction</span>
<span class="no">hws</span> <span class="kw">&lt;-</span> <span class="fl">20</span> <span class="co"># Peak extraction</span>
<span class="no">tol</span> <span class="kw">&lt;-</span> <span class="fl">0.003</span> <span class="co"># Peak binning</span></pre></body></html></div>
</div>
<div id="initial-screening" class="section level2">
<h2 class="hasAnchor">
<a href="#initial-screening" class="anchor"></a>Initial screening</h2>
<p>Pre-processing methods are not immune to low-quality mass spectra and inadvertently letting them through can severely distort the results. For example, when dealing with a number of data sets we observed both clusters of spurious peaks in some cases and sparse peaks in others which interfere with the determination of common reference peaks for peak alignment and binning. Common low-quality mass spectra are characterised by extremely ripply and indented profiles due to low signal intensities. The low signal intensity causes poor peak formation and poorly resolved peaks. Moreover, ion suppression cases show one or a few abnormally high peaks whereas the other peptide features are pulled down. The package <code>MALDIrppa</code> helps to detect exotic signals by a semi-automatic screening process implemented in the <code>screenSpectra</code> function. It is based on robust scale estimators <span class="citation">(Rousseeuw and Croux 1993)</span> of the derivative mass spectra and median intensities. These components are given a weight to build an atypicality score (<span class="math inline">\(A\)</span>) for each signal, which is then labelled as a suspicious case (<code>failure</code>) if the score falls beyond the estimated tolerance limits (see help for <code>screenSpectra</code> for options available). The atypicality score <span class="math inline">\(A\)</span> is calculated as</p>
<p><span class="math display">\[A = \frac{S^{\lambda}}{(\texttt{med}+1)^{(1-\lambda)/2}},\]</span></p>
<p>where <span class="math inline">\(S\)</span> is the robust scale estimator used (either Rousseeuw’s <span class="math inline">\(Q\)</span> or the well-known median absolute deviance, MAD, estimator are available), <span class="math inline">\(\texttt{med}\)</span> is the median intensity of the raw signal and <span class="math inline">\(\lambda\)</span> distributes the weight given to each of these components (<span class="math inline">\(\lambda = 1/2\)</span> by default).</p>
<p>The <code>screenSpectra</code> function generates a table <code>est.table</code> with individual information for each mass spectrum. The following code illustrates its use with default settings and creates a new <code>scSpectra</code> class object called <code>sc.results</code>. The function can take a matrix of metadata (<code>type</code> in our case) as argument. This is convenient to easily filter problematic cases out in both the spectra and metadata sets for subsequent analysis. Ordinary R methods <code>summary</code> and <code>plot</code> can be applied on <code>scSpectra</code> class objects to obtain a summary table and graphs of the results.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">sc.results</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/screenSpectra.html">screenSpectra</a></span>(<span class="no">spectra</span>, <span class="kw">meta</span> <span class="kw">=</span> <span class="no">type</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">sc.results</span>)</pre></body></html></div>
<pre><code>## (10 first mass spectra) 
##           ID   A score   Class
## 1  160408F21 0.1683850 success
## 2  160408F22 0.1716699 success
## 3  160408F23 0.1887859 success
## 4  160408F24 0.1478541 success
## 5  160408G01 0.1479799 success
## 6  160408G02 0.1734508 success
## 7  160408G03 0.1772491 success
## 8  160408G04 0.1534341 success
## 9  160408G05 0.1716697 success
## 10 160408G06 0.1538116 success
## 
## ----------------------------
## 
## Scale estimator: Q 
## Method: adj.boxplot 
## Threshold: 1.5 
## Limits: [0.1196,0.2942] 
## Deriv. order: 1 
## Lambda: 0.5 
## No. potentially faulty spectra: 19 (4.75 %)</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">sc.results</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="MALDIrppa_vignette_files/figure-html/iniscreening-1.png" width="528" style="display: block; margin: auto;"></p>
<p>Among other features, the method <code>plot</code> allows to visualise the distribution of the atypicality scores (<code>type = "hist"</code>), customise the point labels and interactively display marked spectra for individual visual inspection (<code>type = "casewise"</code>). When <code>label = TRUE</code> it shows the position index of the mass spectra in the data set. It is advisable to visually inspect marked mass spectra, particular borderline cases, and further investigate any pattern. For example, in our illustrative data set we observe that mass spectra number 25 to 28 correspond to a series of technical replicates within the same biological replicates (see e.g. <code>type[20:30,]</code>). Focusing on some of the most extreme <span class="math inline">\(A\)</span> scores, the following figures display mass spectra 28, 29 and 87, which correspond with low signal intensity, ion suppression and flatline mass spectra respectively. Flatline mass spectra are produced when the instrument is unable to collect data that does not meet the minimum resolution and intensity requirements. These low-quality mass spectra are clearly identified using the <span class="math inline">\(A\)</span> score as seen above.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">spectra</span><span class="kw">[[</span><span class="fl">28</span>]])</pre></body></html></div>
<p><img src="MALDIrppa_vignette_files/figure-html/spectraplots10-1.png" width="528" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">spectra</span><span class="kw">[[</span><span class="fl">29</span>]])</pre></body></html></div>
<p><img src="MALDIrppa_vignette_files/figure-html/spectraplots10-2.png" width="528" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">spectra</span><span class="kw">[[</span><span class="fl">87</span>]])</pre></body></html></div>
<p><img src="MALDIrppa_vignette_files/figure-html/spectraplots10-3.png" width="528" style="display: block; margin: auto;"></p>
<p>As an example of the problems caused by faulty, low-quality spectra, the next figure shows the case of two peak profiles originated from the same biological material. It is evident that they are carrying very different information.</p>
<center>
<img src="lqgq_plot.png" alt="Drawing" style="width: 500px;">
</center>
<p>For the purpose of this tutorial we move on by updating the original <code>spectra</code> and <code>type</code> objects with the filtered collection of mass spectra and associated metadata provided by <code>screenSpectra</code> through the <code>fspectra</code> and <code>fmeta</code> elements of the output.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">spectra</span> <span class="kw">&lt;-</span> <span class="no">sc.results</span>$<span class="no">fspectra</span> <span class="co"># Filtered list of mass spectra</span>
<span class="no">type</span> <span class="kw">&lt;-</span> <span class="no">sc.results</span>$<span class="no">fmeta</span> <span class="co"># Filtered metadata</span></pre></body></html></div>
</div>
<div id="denoising-baseline-correction-and-normalisation" class="section level2">
<h2 class="hasAnchor">
<a href="#denoising-baseline-correction-and-normalisation" class="anchor"></a>Denoising, baseline correction and normalisation</h2>
<p>Signal denoising or smoothing is conducted in <code>MALDIrppa</code> by undecimated discrete wavelet transform (UDWT) <span class="citation">(Coombes et al. 2005)</span> using the <code>wavSmooting</code> function. Wavelets adapt well to changes in peak shape and the UDWT produces the same results if the start of the signal is shifted by a few time points (shift-invariance). This prevents from undesirable artifacts into the signal near either end of the spectrum due to large changes in wavelet coefficients as a consequence of small shifts in location. In this case, smoothing is applied on square root-transformed (<code>sqrt</code> function in R) signals using <code>transfIntensity</code> for variance stabilisation. Note that in fact this function allows to apply any sensible user-defined transformation on the signal intensities (see <code>MALDIrppa</code> documentation for an example). For baseline correction, the statistics-sensitive non-linear iterative peak-clipping (SNIP) algorithm as implemented in <code>MALDIquant</code> generates positive intensities and provides better results with high matrix effects (pronounced mode). Regarding signal normalisation, <code>MALDIquant</code> offers a number of sensible methods through the <code>calibrateIntensity</code> function. In particular, probabilistic quotient normalisation (PQN) provides a robust method for scaling spectra, commonly originated from different dilutions, into a common overall concentration <span class="citation">(Dieterie et al. 2006)</span>.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">spectra</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/transfIntensity.html">transfIntensity</a></span>(<span class="no">spectra</span>, <span class="kw">fun</span> <span class="kw">=</span> <span class="no">sqrt</span>)
<span class="no">spectra</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/wavSmoothing.html">wavSmoothing</a></span>(<span class="no">spectra</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"Wavelet"</span>, <span class="kw">thresh.scale</span> <span class="kw">=</span> <span class="no">thScale</span>)
<span class="no">spectra</span> <span class="kw">&lt;-</span> <span class="fu">removeBaseline</span>(<span class="no">spectra</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"SNIP"</span>, <span class="kw">iterations</span> <span class="kw">=</span> <span class="no">ite</span>)
<span class="no">spectra</span> <span class="kw">&lt;-</span> <span class="fu">calibrateIntensity</span>(<span class="no">spectra</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"PQN"</span>)</pre></body></html></div>
<p>Note that in this illustrative pipeline we go on overwriting the <code>MassSpectrum</code> object containing the list of signals as we apply the different operations to produce a clean and corrected collection of signals. Given the usual large size of the data set, this is convenient to save computer memory and speed up the process. However we do not then keep individual backup copies of the spectra at each pre-processing stage. If this is required, simply using the ordinary <code><a href="https://rdrr.io/r/base/save.html">save(objectname, filename)</a></code> line in R after each stage will do. In any case, it is recommended to delete previous versions of the data from R’s workspace as soon as they are saved in order to prevent from memory usage problems.</p>
</div>
<div id="peak-extraction-and-alignment" class="section level2">
<h2 class="hasAnchor">
<a href="#peak-extraction-and-alignment" class="anchor"></a>Peak extraction and alignment</h2>
<p>Using <code>detectPeaks</code> in <code>MALDIquant</code> peak profiles are produced by searching for and extracting local maximum intensities along the corrected MS signals. A noise function is estimated (see <code>estimateNoise</code>) and signal-to-noise ratio (SNR) and mass window size values are set in order to determine peaks corresponding to genuine peptide-induced signals. In particular, peaks are extracted if their intensity is greater than SNR times the estimated noise within a mass <span class="math inline">\(\pm\)</span> half window size interval. Extracted peak lists still require alignment and binning across profiles. This is necessary to correct for slight mass drifts and arrange all peak profiles onto a common range of masses. A number of landmark peaks that are matched a minimum number of times across profiles serve as reference points for alignment, following procedures implemented in <code>MALDIquant</code> (the <code>minFreq</code> argument sets this as a proportion, 80% frequency in our example). Binning is then used to equalise masses of peaks found within a tolerable relative mass deviation range and, hence, assumed to be representatives of the same peptide (the <code>tolerance</code> argument sets this as a proportion, 0.3% deviance in our case). The <code>alignPeaks</code> function in <code>MALDIrppa</code> is a wrapper for a number of <code>MALDIquant</code> functions involved in this stage that simplifies the workflow into a single line of code. Moreover, <code>alignPeaks</code> implements an additional binning round which we found useful to deal with misalignment issues that may be still present after putting the peak profiles through default <code>MALDIquant</code> alignment and binning procedures. We faced this situation when working with mass spectra over a particularly high-resolution mass range.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">peaks</span> <span class="kw">&lt;-</span> <span class="fu">detectPeaks</span>(<span class="no">spectra</span>, <span class="kw">SNR</span> <span class="kw">=</span> <span class="no">SigNoi</span>, <span class="kw">halfWindowSize</span> <span class="kw">=</span> <span class="no">hws</span>)
<span class="no">peaks</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/alignPeaks.html">alignPeaks</a></span>(<span class="no">peaks</span>, <span class="kw">minFreq</span> <span class="kw">=</span> <span class="fl">0.8</span>, <span class="kw">tolerance</span> <span class="kw">=</span> <span class="no">tol</span>)</pre></body></html></div>
<p>It is important to keep exploring the data throughout the pre-processing workflow in order to identify potential issues or unexpected results. Analogously to <code>summarySpectra</code> used above, <code>summaryPeaks</code> produces a table of summary statistics from the peak profiles. The following line of code shows for example details for the first 10 peak profiles.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="../reference/summaryPeaks.html">summaryPeaks</a></span>(<span class="no">peaks</span>[<span class="fl">1</span>:<span class="fl">10</span>])</pre></body></html></div>
<pre><code>##    ID No.MZ   Min.MZ   Max.MZ Min.Int Mean.Int Std.Int Med.Int MAD.Int Max.Int Min.SNR Mean.SNR Std.SNR Med.SNR MAD.SNR Max.SNR
## 1   1    14 3174.229 10499.25   1e-04    3e-04   2e-04   4e-04   2e-04   7e-04  2.1462   5.1254  2.3576  5.3609  2.6891 10.5442
## 2   2    16 3160.482 10499.25   1e-04    4e-04   2e-04   4e-04   2e-04   8e-04  2.0941   5.7455  2.9227  6.0530  3.3212 12.7601
## 3   3    13 3174.229 10499.25   1e-04    4e-04   2e-04   4e-04   2e-04   8e-04  2.0678   5.1162  2.3704  5.4186  2.3103 10.4369
## 4   4    14 3174.229 10499.25   2e-04    3e-04   2e-04   4e-04   2e-04   7e-04  2.4580   5.5974  2.6155  5.7753  3.1150 11.4999
## 5   5    14 2835.022 10499.25   1e-04    3e-04   2e-04   3e-04   2e-04   6e-04  2.1403   5.7576  2.9917  4.8636  3.5934 11.2820
## 6   6    17 2835.022 12181.44   1e-04    3e-04   2e-04   2e-04   2e-04   7e-04  2.3676   6.3853  3.8149  4.7608  3.3397 13.1390
## 7   7    15 3128.287 10499.25   1e-04    4e-04   2e-04   3e-04   3e-04   8e-04  2.0799   6.2812  3.5367  5.1705  4.3113 13.3856
## 8   8    16 2835.022 10499.25   1e-04    3e-04   2e-04   3e-04   2e-04   7e-04  2.2752   5.5075  2.9669  4.6241  3.3451 11.5115
## 9   9    14 3174.229 10499.25   2e-04    3e-04   1e-04   3e-04   2e-04   6e-04  2.5949   5.6028  2.5182  5.7975  3.2635 10.6196
## 10 10    19 2835.022 12181.44   1e-04    3e-04   2e-04   2e-04   1e-04   7e-04  2.3271   6.3129  3.9666  4.6307  3.0533 15.1524</code></pre>
<p>The function <code>countPeaks</code> directly counts the number of peaks in each profile. Either an obvious excess or scarcity of peaks in a profile may deserve a closer look. The following code produces the list of counts and plots them for overall comparison using the profile number as label.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">cP</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/countPeaks.html">countPeaks</a></span>(<span class="no">peaks</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">cP</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"n"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="no">cP</span>, <span class="kw">label</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">cP</span>))</pre></body></html></div>
<p><img src="MALDIrppa_vignette_files/figure-html/npeaks-1.png" width="528" style="display: block; margin: auto;"></p>
<p>At this point we have essentially concluded data pre-processing and it is then convenient to save a backup of the peak profiles and associate metadata.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span>(<span class="no">peaks</span>, <span class="kw">file</span> <span class="kw">=</span> <span class="st">"peaks.pp.Rdata"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span>(<span class="no">type</span>, <span class="kw">file</span> <span class="kw">=</span> <span class="st">"type.pp.Rdata"</span>)</pre></body></html></div>
<p>The <code>peakPatterns</code> function is useful to visualise and explore the resulting peak patterns. It displays peak presence (coloured cells) and absence (blank cells). The barplot on the top reflects the frequency of a peak across replicates. The function can be applied either on a list of <code>MassPeaks</code> objects, like our <code>peaks</code>, or on an intensity matrix. Any remaining data artifacts or hints about peak features differentiating e.g. strains should be evident here. This display can also be useful to check for homogeneity of replicates from an isolate by simply subsetting the replicates within that same isolate (see <code><a href="../reference/peakPatterns.html">?peakPatterns</a></code> for more details and examples).</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu"><a href="../reference/peakPatterns.html">peakPatterns</a></span>(<span class="no">peaks</span>)</pre></body></html></div>
<p><img src="MALDIrppa_vignette_files/figure-html/intmatrix-1.png" width="528" style="display: block; margin: auto;"></p>
</div>
<div id="outliers-filtering-and-merging-replicates" class="section level2">
<h2 class="hasAnchor">
<a href="#outliers-filtering-and-merging-replicates" class="anchor"></a>Outliers, filtering and merging replicates</h2>
<p>Note that due to process disturbances, chemical contamination or human errors for example, some peak profiles may exhibit patterns beyond anticipated biological variability at a level at which certain homogeneity is expected, say at biological and technical replication level. This is not related to low-quality, data acquisition problems as before, the signals are correct, it is instead a case of outlying peak profiles characterised for its large distance from the main mass of data points. They usually include relevant information about certain artifacts or substructures in the data, although in practice it is not always straightforward to find the underlying reasons for their deviating behaviour. As it is well-known in general data modelling, outliers can heavily influence e.g. correlations, fitting in regression models or similarity measures used by classification algorithms. Outliers in our context are often sought after on a peptide-wise basis <span class="citation">(Erhard and Zimmer 2012)</span>, but we adhere to the idea that peaks of outlying profiles do not necessarily stand out individually. They can simply show an atypical relative pattern with combinations of peaks that occur very rarely. Working within a robust statistics framework we consider the multidimensional and inter-dependent structure of the data for outlier detection <span class="citation">(Rocke and Woodruff 1996; Maronna, Martin, and Yohai 2006)</span>. In particular, we adapt the multivariate outlier detection method proposed by <span class="citation">Filzmoser, Garrett, and Reimann (2005)</span> (adaptive outlier detection, AOD) which allows the boundaries for a peak profile to be considered as potential outlier to be adjusted to the number of replicates involved. By using a multidimensional scaling (MDS) coordinate projections of the data, the approach can be applied at different levels of replication and on varied data formats, e.g. either peak intensities or peak presence/absence binary data. This customised procedure is implemented in the <code>detectOutliers</code> function. It generates a logical vector flagging out potential outlying replicates at a given level of aggregation, e.g. at isolate level as set using the argument <code>by</code> in the following line of code. This can be used to filter outliers out if that is judged adequate. The argument <code>binary</code> facilitates the use of the method on either peak presence/absence profiles (<code>binary = TRUE</code>) or raw peak intensity profiles (<code>binary = FALSE</code>).</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">out</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/detectOutliers.html">detectOutliers</a></span>(<span class="no">peaks</span>, <span class="no">type</span>$<span class="no">Isolate</span>, <span class="kw">binary</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>The following figure illustrates the result from a collection of 20 replicates from the same bacterial isolate. Note that the method was applied on peak presence/absence profiles. Quantile ellipsoids based on the robust Mahalanobis distance determine outlierness thresholds. The graph displays ellipsoids at increasing quantiles, including the adjusted quantile ellipsoid produced by the AOD method. The profiles, projected onto a 2-dimensional space by MDS, falling beyond the adjusted quantile boundaries were marked as potential outliers. In this case there is evidence of a set of 4 technical replicates within a biological replicate that were probably contaminated samples (distant points located on the right-hand side of the graph).</p>
<center>
<img src="od_plot.png" alt="Drawing" style="width: 500px;">
</center>
<p>The identification of outliers can be sometimes a primary purpose of data analysis, as they might reveal hidden structures in the data of biological interest. Other times outliers are a nuisance resulting from errors or contamination, and it might be then convenient to discard or downweight them prior to the actual data modelling. <code>MALDIrppa</code> provides a exploratory tool to inform of samples that are considerably different from the majority. The final decision about their biological meaning and how to deal with them will depend on the assessment of the researcher. To simplify the exposition we here proceed by discarding outlying peak profiles.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">peaks.clean</span> <span class="kw">&lt;-</span> <span class="no">peaks</span>[<span class="no">out</span>[,<span class="fl">2</span>] <span class="kw">==</span> <span class="fl">FALSE</span>] <span class="co"># Discard outlying peak profiles</span>
<span class="no">type.clean</span> <span class="kw">&lt;-</span> <span class="no">type</span>[<span class="no">out</span>[,<span class="fl">2</span>] <span class="kw">==</span> <span class="fl">FALSE</span>,]  <span class="co"># and corresponding metadata</span></pre></body></html></div>
<p>The inherent nature of the data implies that pre-processed data may still contain peaks only reflecting noise and probably irrelevant signals. <code>MALDIquant</code> facilitates further cleaning by eliminating rare peaks across replicates using the <code>filterPeaks</code> function. This is done by setting a minimum frequency (<code>minFreq</code> argument) for a peak at a desired level of aggregation (isolate level in our case). Note that in <code>MALDIquant</code> the argument <code>labels</code> is used to set the grouping factor. The following line discards peaks occurring in less than 25% replicates within isolate.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">peaks.clean.f</span> <span class="kw">&lt;-</span> <span class="fu">filterPeaks</span>(<span class="no">peaks.clean</span>, <span class="kw">minFreq</span> <span class="kw">=</span> <span class="fl">0.25</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="no">type.clean</span>$<span class="no">Isolate</span>)</pre></body></html></div>
<p>We can compare these filtered peak profiles with the more noisy raw peaks profiles above using <code>peakPatterns</code>.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="../reference/peakPatterns.html">peakPatterns</a></span>(<span class="no">peaks.clean.f</span>)</pre></body></html></div>
<p><img src="MALDIrppa_vignette_files/figure-html/intmatrix2-1.png" width="528" style="display: block; margin: auto;"></p>
<p>Finally replicates are typically merged to obtain a single composite representative peak profile by isolate. In accordance with our robust approach we propose to do this by computing the median peak profile across replicates within isolate. This contributes to downgrade the influence of extreme peak intensities on the pooled estimates. The associated metadata should be also aggregated in the same way. Here we use the <code>aggregate</code> function available in R base, but other alternatives are available through several R packages. Note that we specify <code>FUN = length</code> simply as a <em>trick</em>, as we are not really interested in any summary of the metadata. Moreover, we only keep the information about the isolate, which is in the first column of the resulting table (note that in this illustrative example <code>type.clean.fm</code> becomes a vector after this).</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">peaks.clean.fm</span> <span class="kw">&lt;-</span> <span class="fu">mergeMassPeaks</span>(<span class="no">peaks.clean.f</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="no">type.clean</span>$<span class="no">Isolate</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"median"</span>)
<span class="no">type.clean.fm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span>(<span class="no">type.clean</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">Isolate</span> <span class="kw">=</span> <span class="no">type.clean</span>$<span class="no">Isolate</span>), <span class="kw">FUN</span> <span class="kw">=</span> <span class="no">length</span>)[,<span class="fl">1</span>]</pre></body></html></div>
<p>Statistical and machine learning methods typically work on data matrices. The <code>intensityMatrix</code> function in <code>MALDIquant</code> converts a list of <code>MassPeaks</code> objects into a matrix. The replicates or composite profiles are arranged by rows, and the columns correspond with the list of transformed intensities along the common <span class="math inline">\(m/z\)</span> range determined after peak alignment and binning. When a peak is not present at a certain mass position the cell is given a <code>NA</code> value by default. The final intensity matrix is obtained as follows.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">int.clean.fm</span> <span class="kw">&lt;-</span> <span class="fu">intensityMatrix</span>(<span class="no">peaks.clean.fm</span>)</pre></body></html></div>
<p>The intensity matrix and associated metadata can be stored in different formats to facilitate its use as input in, for example, specialised bioinformatics software packages. The following code uses simple functions to save them as CSV text files and also in the popular NEXUS format, which includes options to save either peak intensities or binary patterns and to add peak weights among others (see <code><a href="../reference/writeIntensity.html">?writeIntensity</a></code> help files for more details). The argument <code>labels</code> can be used to add labels to the pre-processed samples. For this example data set, we use the spectra IDs stored in <code>type.clean.fm</code>, which became a vector containing only the IDs in a previous stage. Thus, using the <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> operator for matrices is not required to select the corresponding ID column here.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="../reference/writeMetadata.html">writeMetadata</a></span>(<span class="no">type.clean.fm</span>, <span class="kw">filename</span> <span class="kw">=</span> <span class="st">"type.clean.fm"</span>, <span class="kw">format</span> <span class="kw">=</span> <span class="st">"csv"</span>) <span class="co"># save metadata</span>
<span class="fu"><a href="../reference/writeIntensity.html">writeIntensity</a></span>(<span class="no">int.clean.fm</span>, <span class="kw">filename</span> <span class="kw">=</span> <span class="st">"int.clean.fm"</span>, <span class="kw">format</span> <span class="kw">=</span> <span class="st">"csv"</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="no">type.clean.fm</span>)
<span class="fu"><a href="../reference/writeIntensity.html">writeIntensity</a></span>(<span class="no">int.clean.fm</span>, <span class="kw">filename</span> <span class="kw">=</span> <span class="st">"int.clean.fm"</span>, <span class="kw">format</span> <span class="kw">=</span> <span class="st">"NEXUS"</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="no">type.clean.fm</span>)</pre></body></html></div>
<p>The argument <code>binary</code> can be used to export data in binary (peak presence/absence) format. Note that this is internally turned <code>TRUE</code> whenever the NEXUS or FASTA format is chosen. As an example, the next figure shows a phylogenetic network contructed using the NeighborNet algorithm implemented in the SplitsTree software package (<a href="http://www.splitstree.org" class="uri">http://www.splitstree.org</a>) directly from the NEXUS file generated above by <code>MALDIrppa</code>.</p>
<center>
<img src="phylo_net.png" alt="Drawing" style="width: 600px;">
</center>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Albrethsen2007">
<p>Albrethsen, Jakob. 2007. “Reproducibility in protein profiling by MALDI-TOF mass spectrometry.” <em>Clinical Chemistry</em> 53 (5): 852–8. <a href="https://doi.org/10.1373/clinchem.2006.082644">https://doi.org/10.1373/clinchem.2006.082644</a>.</p>
</div>
<div id="ref-Coombes2005">
<p>Coombes, Kevin R, Spiridon Tsavachidis, Jeffrey S Morris, Keith a Baggerly, Mien-Chie Hung, and Henry M Kuerer. 2005. “Improved peak detection and quantification of mass spectrometry data acquired from surface-enhanced laser desorption and ionization by denoising spectra with the undecimated discrete wavelet transform.” <em>Proteomics</em> 5 (16): 4107–17. <a href="https://doi.org/10.1002/pmic.200401261">https://doi.org/10.1002/pmic.200401261</a>.</p>
</div>
<div id="ref-Diet2006">
<p>Dieterie, Frank, Alfred Ross, Götz Schlotterbeck, and Hans Senn. 2006. “Probabilistic Quotient Normalization as Robust Method to Account for Dilution of Complex Biological Mixtures. Application in 1H NMR Metabonomics.” <em>Analytical Chemistry</em> 78 (13): 4281–90. <a href="https://doi.org/10.1021/ac051632c">https://doi.org/10.1021/ac051632c</a>.</p>
</div>
<div id="ref-Erhard2012">
<p>Erhard, Florian, and Ralf Zimmer. 2012. “Detecting outlier peptides in quantitative high-throughput mass spectrometry data.” <em>Journal of Proteomics</em> 75 (11): 3230–9. <a href="https://doi.org/10.1016/j.jprot.2012.03.032">https://doi.org/10.1016/j.jprot.2012.03.032</a>.</p>
</div>
<div id="ref-Filzmoser2005">
<p>Filzmoser, Peter, Robert G Garrett, and Clemens Reimann. 2005. “Multivariate outlier detection in exploration geochemistry.” <em>Computers &amp; Geosciences</em> 31: 579–87. <a href="https://doi.org/10.1016/j.cageo.2004.11.013">https://doi.org/10.1016/j.cageo.2004.11.013</a>.</p>
</div>
<div id="ref-Gibb2012">
<p>Gibb, Sebastian, and Korbinian Strimmer. 2012. “MALDIquant: a versatile R package for the analysis of mass spectrometry data.” <em>Bioinformatics</em> 28 (17): 2270–1. <a href="https://doi.org/10.1093/bioinformatics/bts447">https://doi.org/10.1093/bioinformatics/bts447</a>.</p>
</div>
<div id="ref-Maronna2006">
<p>Maronna, Ricardo A., R. Douglas Martin, and Vı́ctor J. Yohai. 2006. <em>Robust Statistics: Theory and Methods</em>. Wiley Series in Probability and Statistics. Chichester, UK: John Wiley &amp; Sons, Ltd. <a href="https://doi.org/10.1002/0470010940">https://doi.org/10.1002/0470010940</a>.</p>
</div>
<div id="ref-Mclean2017">
<p>Mclean, K., J. Palarea-Albaladejo, C. G. Currie, L. H. J. Imrie, E. D. T. Manson, D. Fraser-Pitt, F. Wright, et al. 2017. “Whole cell MALDI mass spectrometry provides a rapid approach to differentiation of STEC strains.” <em>(Submitted)</em>.</p>
</div>
<div id="ref-Rocke1996">
<p>Rocke, David M, and David L Woodruff. 1996. “Identification of Outliers in Multivariate Data.” 91 (435): 1047–61.</p>
</div>
<div id="ref-Rousseeuw1993">
<p>Rousseeuw, Peter J., and Christophe Croux. 1993. “Alternatives to the median absolute deviation.” <em>Journal of the American Statistical Association</em> 88 (424): 1273–83. <a href="https://amstat.tandfonline.com/doi/abs/10.1080/01621459.1993.10476408">https://amstat.tandfonline.com/doi/abs/10.1080/01621459.1993.10476408</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Javier Palarea-Albaladejo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
